sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -142,6 +142,8 @@ object TypeCoercion {
     case (n: DecimalType, s: StringType) => Some(DoubleType)
     case (s: StringType, n: DecimalType) => Some(DoubleType)
 
+    case (l: StringType, r: IntegerType) => Some(DoubleType)
```

Comment:
shall we apply it to all numeric types?

since we don't know the value of the string, using double seems safest to avoid overflow.

Reply:
Using double has other issues:
```scala
scala> spark.sql("select cast('9223372036854775806' as double) = cast(cast('9223372036854775805' as bigint) as double) as toDouble").show
+--------+
|toDouble|
+--------+
|    true|
+--------+
```

Reply:
![image](https://user-images.githubusercontent.com/6757692/72146292-17e39f00-33d7-11ea-8f22-8b5ec60a7a5a.png)

Reply:
![image](https://user-images.githubusercontent.com/6757692/72146437-62fdb200-33d7-11ea-9063-fa6c59cb63df.png)

Reply:
![image](https://user-images.githubusercontent.com/6757692/72146648-d9021900-33d7-11ea-8d92-0e5137619f35.png)

Reply:
Just attach a unit test:
![image](https://user-images.githubusercontent.com/6757692/72147093-cc31f500-33d8-11ea-8baf-c2cd0d80d92b.png)

Reply:
It seems that for a double value, if its value exceeds 10^15, the value comparison is not accurate, because DoubleType's scale is at most 15.

Reply:
It seems that when comparing StringType and IntegerType, Decimal(38, 0) is a good option.

Reply:
how about `'1.2' = 1`?

Reply:
Hm, It seems Decimal(38, 1) is more appreciate.

Reply:
Decimal(38, 1) doesn't work for `'1.23' = 1`.

I don't think there is perfect solution here. Maybe we should follow pgsql and force users to add cast, when ansi mode is on.

Reply:
Yes, I think we should follow pgsql when ansi is enabled.